docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,322 @@
+severe impact on system stability:
+The vertical scaling stabilization algorithm used by HVPA revolves around a 'MinChange' concept, essentially suppressing
+vertical scaling until the recommendation differs from the current allocation (request) by said amount.
+Alone, this concept is functionally incorrect when applied to scale-up: scaling is suppressed in a state where VPA is
+indicating that the workload is starving for resources. To mitigate this problem, the existing design relies on
+resource headroom existing on the node, which can temporarily accommodate the starving workload's needs. In the field,
+this approach has been repeatedly observed to cause problems for etcd, for which a particularly high value of MinChange
+is used, as frequent evictions are highly undesirable. As an illustration, take a pod with MinChange for CPU set to
+1000 millicores, current CPU request at X millicores, actual pod need at X+2000, and available node headroom at 500.
+What happens is that actual pod usage extends as far as headroom allows, reaching X+500 and is starving for extra 1500
+millicores. Based on pod usage at X+500, VPA recommends e.g. X+900. This recommendation is below the MinChange threshold
+of X+1000. HVPA does not act, and the pod is permanently stuck starving for 1500 millicores of CPU.
+
+Lastly, HVPA applies additive model to the interaction between HPA and VPA, with respect to provided compute power, but
+that interaction is actually multiplicative. As an extreme example, take a workload running at one replica with one core.
+Then consumption changes to 5 cores, leading HPA to recommend 5 single-core replicas - a change of +4 replicas. VPA
+suggests adding 4 cores to the existing single replica. A HVPA configured at 50% HPA + 50% VPA, would apply half of
+each recommendation, +2 cores and +2 replicas, resulting in (1 + 2) replicas * (1 + 2) cores. Failing to account for
+the multiplicative interaction between changes, HVPA provides a total of 9 cores to a workload which only needs 5. 
+
+### Other Benefits to Replacing the Existing Solution
+- HVPA applies recommendations via direct edits to the pod objects. This approach requires nonobvious special handling
+  by Gardener code outside HVPA, which is error-prone.
+
+### Goals
+- Improve ShootKapi autoscaling stability by eliminating the stability issues existing in the current HVPA-based solution
+- Improve ShootKapi operational efficiency at low application load levels, by avoiding excessive replicas driven by
+  HVPA's policy for early horizontal scale-out
+- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions
+- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi
+- Reduce Gardener custom code base and ongoing maintenance effort
+
+### Non-Goals
+- There are other components currently scaled via HVPA. Changing the autoscaling mode for these is outside the scope
+  of this GEP
```

Comment:
Why? What's the plan with these other components? Will there be yet another scaling strategy for them? I think it makes sense to reduce complexity and also consider them as part of this GEP so that the goal of "dropping HVPA" can be achieved faster.

Reply:
We're also planning to get rid of HVPA usage for the other components, but the approach will be entirely different. kube-apiserver is the only component using HVPA to combine VPA and HPA. All other components use HVPA to allow for tuning VPA configuration beyond what open-source VPA allows.
Therefore, we decided to keep this out of this KEP – it will be a separate KEP, covering a separate problem with a separate solution. Does this make sense?

Reply:
Yes, it does. Can you please put exactly this explanation into the document then? I think this might definitely be of interested to other readers :)

Reply:
Done. Will be available with next push.